src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java

```diff
@@ -0,0 +1,325 @@
+        return map;
+    }
+
+    /**
+     * Uses {@link UserService} to create a user in the database
+     * @param u User to add to database
+     */
+    private void commitUser(User u) {
+        u.setSystemRole(Role.ROLE_USER);
+        try {
+            ldapUserRevision = true;
+            userService.create(u);
+        } catch (EntityExistsException e) {
+            logger.error("User being created already exists: " + e);
+            throw e;
+        } catch (ConstraintViolationException e) {
+            logger.error("Fields in User are incompatible with local database constraints: " + e);
+            throw e;
+        } finally {
+            ldapUserRevision = false;
+        }
+    }
+
+    /**
+     * Uses {@link UserService} to update a user in the database
+     * @param u User to update in database
+     * @param updatedProperties Map of userproperties to update
+     */
+    private void updateUser(User u, Map<String, Object> updatedProperties) {
+        try {
+            ldapUserRevision = true;
+            userService.updateFields(u.getId(), updatedProperties);
+        } catch (ConstraintViolationException e) {
+            logger.error("Fields in User are incompatible with local database constraints: " + e);
+            throw e;
+        } catch (EntityNotFoundException | EntityExistsException | InvalidPropertyException e) {
+            // This should be unreachable, but is here for safety
+            logger.error("IRIDA was unable to update user from LDAP/ADLAP fields: " + e);
+            throw e;
+        } finally {
+            ldapUserRevision = false;
+        }
+    }
+
+    /**
+     * Fetches attribute from Ldap/adLdap service
+     * @param dirContextOperations Ldap/adLdap context
+     * @param field String as defined in the Ldap/adLdap server
+     * @param required Boolean, if required and not found, an {@link IridaLdapAuthenticationException} will be thrown
+     * @param fallback String, default value
+     * @return String value fetched from Ldap/adLdap service
+     */
+    private String getAttribute( DirContextOperations dirContextOperations, String field, boolean required, @NotNull String fallback) {
+        Locale locale = LocaleContextHolder.getLocale();
+
+        String res = dirContextOperations.getStringAttribute(field);
+
+        if (res==null || res.equals("")) {
+            if (required) {
+                String ldap_error4 = messageSource.getMessage(
+                        "LoginPage.ldap_error.description_4", null, locale);
+                logger.error(ldap_error4);
+                throw new IridaLdapAuthenticationException(ldap_error4, 4);
+            }
+            else {
+                res = fallback;
+            }
+        }
+        return res;
+    }
+
+    /**
+     * Generates a random password that adheres to our password requirements
+     * @return String
+     */
+    public String generateCommonLangPassword() {
```

Comment:
Instead of having to generate a random password wouldn't it make more sense to create another entity i.e. `DomainUser` that allows Null for password field. Something like how we have `CoverageQCEntry` that is on top of `QCEntry` and has the `DTYPE` column. That might complicate password checking for non `DomainUser` accounts as we would not have the db NULL constraint. I guess this could be tackled in your other PR where you allow local and domain users.

Reply:
That's likely the better way to do it yeah. I'll work on that as part of the next PR as this one already is a bit bloated.